Framework/Kernel/src/Unit.cpp

```diff
@@ -644,56 +644,80 @@ void dSpacing::init() {
                              "has been initialized.");
-  return difa * x * x + difc * x + tzero;
+  if (difa == 0.)
+    return difc * x + tzero;
+  else
+    return difa * x * x + difc * x + tzero;
 }
 
+/**
+ * DIFA * d^2 + DIFC * d + T0 - TOF = 0
+ *
+ * Use the citardauq formula to solve quadratic in order to minimise loss of precision. citarqauq (quadratic spelled
+ * backwards) is an alternate formulation of the quadratic formula. DIFC and sqrt term are often similar and the
+ * "classic" quadratic formula involves calculating their difference in the numerator
+ *
+ *               2*(T0 - TOF)                                            (T0 - TOF)
+ * d = -------------------------------------------  =  ---------------------------------------------------
+ *     -DIFC -+ SQRT(DIFC^2 - 4*DIFA*(T0 - TOF))       0.5 * DIFC (-1 -+ SQRT(1 - 4*DIFA*(T0 - TOF)/DIFC^2)
+ *
+ * the variables in this formulation are the same as the quadratic formula
+ * a = difa      square term
+ * b = DIFC      linear term - assumed to be positive
+ * c = T0 - TOF  constant term
+ */
 double dSpacing::singleFromTOF(const double tof) const {
-  // DIFA * d^2 + DIFC * d + T0 - TOF = 0
-
-  // Use the citardauq formula to solve quadratic in order to minimise loss of precision. DIFC and sqrt term are often
-  // similar and the "classic" quadratic formula involves calculating their difference in the numerator
-
-  //               2*(T0 - TOF)                                            (T0 - TOF)
-  // d = -------------------------------------------  =  ---------------------------------------------------
-  //     -DIFC -+ SQRT(DIFC^2 - 4*DIFA*(T0 - TOF))       0.5 * DIFC (-1 -+ SQRT(1 - 4*DIFA*(T0 - TOF)/DIFC^2)
-
   // dealing with various edge cases
   if (!isInitialized())
     throw std::runtime_error("dSpacingBase::singleFromTOF called before object "
                              "has been initialized.");
   if (!toDSpacingError.empty())
     throw std::runtime_error(toDSpacingError);
-  double c = tzero - tof;
+
+  // force the assumption that difc is positive in debug builds
+  assert(difc > 0.);
+
   // handle special cases first...
-  // citardauq formula hides non-zero root if c=0
+
+  // don't need to solve a quadratic when difa==0
+  if (difa == 0.)
+    return (tof - tzero) / difc;
+
+  // citardauq formula hides non-zero root if tof==tzero
   if (tof == tzero) {
-    if (difa < 0)
+    if (difa < 0.)
       return -difc / difa;
     else
-      return 0;
+      return 0.;
   }
-  double sqrtTerm = 1 - 4 * difa * c / (difc * difc);
-  if (sqrtTerm < 0) {
-    throw std::runtime_error("Cannot convert to d spacing. Quadratic doesn't have real roots");
-  }
-  if ((difa > 0) && (c > 0)) {
+  // non-physical result
+  if ((difa > 0.) && (tzero > tof)) {
     throw std::runtime_error("Cannot convert to d spacing. Quadratic doesn't "
                              "have a positive root");
   }
+
+  // general equation
+  const double linearTerm = tzero - tof;
```

Comment:
The check for `difa==0` will be exercised whenever people convert from d-spacing using the instrument geometry. This trades a `std::sqrt(0)` for an `if`. Since `difa==0` simplifies to solving a simple linear equation, I didn't see a reason to add in all the extra math. This is analogous to picking the correct root of the quadratic based on the sign of `difa` rather than always solving for both and seeing which one is physical.

"constant" is likely to be misinterpreted, but I see your point. My bigger concern what having a variable named "c" in the neighborhood of a concept called "DIFC" which tripped me up. I'm happy with an name that make me think less. What if we think of the terms as part of an equation of a generalized polynomial `0=a0 + a1 * x + a2 * x^2`?

Reply:
More on your suggetion above for expanding the error message to check for `(difa >=0) && (c > 0)`. The error message would be misleading for the case of `difa==0` since people wouldn't understand why we're solving a quadratic for a "simple" linear conversion.

Reply:
I guess we'd need to modify the error message here if we did  `(difa >=0) && (c > 0)`. Perhaps a message saying it's not possible to return a positive d spacing is more user friendly even in the quadratic case